lib/widgets/actions.dart

```diff
@@ -29,8 +29,31 @@ abstract final class ZulipAction {
   ///
   /// This is mostly a wrapper around [updateMessageFlagsStartingFromAnchor];
   /// for details on the UI feedback, see there.
+  ///
+  /// A confirmation dialog is shown if the narrow is a non-conversation narrow.
   static Future<void> markNarrowAsRead(BuildContext context, Narrow narrow) async {
     final zulipLocalizations = ZulipLocalizations.of(context);
+    final store = PerAccountStoreWidget.of(context);
+
+    final shouldShowConfirmationDialog = switch (narrow) {
+      CombinedFeedNarrow()
+        || MentionsNarrow()
+        || StarredMessagesNarrow()
+        || KeywordSearchNarrow()
+        || ChannelNarrow() => true,
+      DmNarrow() || TopicNarrow() => false,
+    };
+
+    if (shouldShowConfirmationDialog) {
+      final unreadCount = store.unreads.countInNarrow(narrow);
+      final didConfirm = showSuggestedActionDialog(context: context,
+        title: zulipLocalizations.markAllAsReadConfirmationDialogTitle,
+        message: zulipLocalizations.markAllAsReadConfirmationDialogMessage(unreadCount),
+        actionButtonText: zulipLocalizations.markAllAsReadConfirmationDialogConfirmButton);
+
+      if (await didConfirm.result != true) return;
```

Comment:
nit: these are tightly related; join visually as one stanza:
```suggestion
        actionButtonText: zulipLocalizations.markAllAsReadConfirmationDialogConfirmButton);
      if (await didConfirm.result != true) return;
```